### PR TITLE
puppetboard/app.py: Optimizing the inventory page

### PR DESCRIPTION
**Pull request by corey-hammerton**
Instead of querying the ```/facts``` endpoint in PuppetDB for every fact that it knows about only
query it for the names that we want. The result is a
much smaller result set that is more noticeable on environments with
large numbers of nodes and/or facts.

On one environment with > 2500 fact-names and 35 nodes a curl test
to the facts endpoint without any query string returned in ~3.5 seconds.
Ran another curl test to the facts endpoint with the query string that
the code in this commit creates, the command finished in ~0.040 seconds.

This is an update for the functionality that was introduced in https://github.com/puppet-community/puppetboard/pull/135

### CODE REVIEW

**Comment by corey-hammerton** [comment]
On the same environment the inventory page renders in >= 7.5 <= 8.5 seconds. With this code patch applied it renders in 0.4 seconds

**Comment by daenney** [comment]
:tired_face:. There's no good reason why querying bare `/facts` should be slower than querying `/facts` with `?query` containing all the facts. I have no idea why this happens but IMHO this should be raised upstream as a bug.

On a different track, it would be much more beneficial to fix this in pypuppetdb instead of patching around it in Puppetboard.

**Comment by corey-hammerton** [comment]
@daenney This code patch only queries the /facts endpoint for the facts and values that we want. 

As I mentioned in the description (this is ONLY with curl, not pypuppetdb)

On one environment with > 2500 fact-names and 35 nodes a curl test
to the facts endpoint without any query string returned in ~3.5 seconds.
Ran another curl test to the facts endpoint with the query string that
the code in this commit creates, the command finished in ~0.040 seconds.

**Comment by corey-hammerton** [comment]
Curl test without a query string (output redirected to /dev/null):
```
$ time curl localhost:8080/v3/facts > /dev/null

real    0m2.775s
user    0m0.023s
sys     0m0.013s
```

Curl test with a query string (output redirected to /dev/null):
```
$ time curl localhost:8080/v3/facts -XGET -d 'query=["or", ["=", "name", "puppetversion"], ["=", "name", "kernelrelease"], ["=", "name", "fqdn"], ["=", "name", "ipaddress"], ["=", "name", "architecture"], ["=", "name", "operatingsystemrelease"]]' > /dev/null

real    0m0.039s
user    0m0.020s
sys     0m0.004s
```

**Comment by daenney** [comment]
Yeah ok, but this sentence is confusing:

> Instead of querying PuppetDB for every fact that it knows about only querying it for the fact names that it knows about.

I believe that what you meant to say is "only querying it for the fact names that we want to know about", because PuppetDB knows all fact names meaning it would be the same thing as querying a bare `/facts`.

**Comment by corey-hammerton** [comment]
I'll reword it to something like:

Instead of querying the ```/facts``` endpoint in PuppetDB for every fact that it knows about only query it for the names that we want.

**Comment by corey-hammerton** [comment]
Do you think this PR can be merged?

**Comment by corey-hammerton** [comment]
@daenney and @nibalizer, I want to push a 0.0.5 release before we accept the work done for PuppetDB 3.x and Puppet 4. It would be nice to include this as a minor optimization.

**Comment by nibalizer** [comment]
Woo code go fast.
